File: backend/app/analysis/fact_extraction.py
"""Fact extraction from signal text."""

from __future__ import annotations

from pydantic import BaseModel, Field

from app.llm.models import LLMRequest
from app.llm.prompts import build_fact_extraction_prompt
from app.models.schemas import Fact, FactCategory


class FactExtractionResult(BaseModel):
    """Result of fact extraction."""

    facts: list[Fact] = Field(default_factory=list)


async def extract_facts(
    text: str, provider_name: str = "openai", model: str | None = None
) -> FactExtractionResult:
    """Extract structured facts from signal text."""
    from app.llm.provider import get_provider

    provider = get_provider(provider_name)
    messages = build_fact_extraction_prompt(text)

    request = LLMRequest(messages=messages, model=model, temperature=0.3)
    result = await provider.complete_structured(request, FactExtractionResult)

    return result

File: src/lib/ai/fact-extraction.ts
/**
 * Fact extraction from signal text.
 * Translated from backend/app/analysis/fact_extraction.py
 */

import { z } from "zod";
import { logger } from "@/lib/logger";
import { getProviderWithFailover } from "./provider";
import { buildFactExtractionPrompt } from "./prompts";
import { FactExtractionResultSchema, type FactExtractionResult } from "./types";
import type { ProviderName } from "./provider";
import type { LLMMessage } from "./types";

export async function extractFacts(
  text: string,
  providerName: ProviderName = "openai",
  model?: string
): Promise<FactExtractionResult> {
  const { provider } = getProviderWithFailover(providerName);
  const messages = buildFactExtractionPrompt(text);

  logger.debug("analysis.fact_extraction.start", {
    provider: providerName,
    text_length: text.length,
  });

  const result = await provider.completeStructured(
    messages,
    FactExtractionResultSchema,
    { model, temperature: 0.3 }
  );

  logger.info("analysis.fact_extraction.complete", {
    provider: providerName,
    facts_count: result.facts.length,
  });

  return result;
}

export async function extractFactsWithPrompt<T extends z.ZodTypeAny>(
  messages: LLMMessage[],
  schema: T,
  providerName: ProviderName = "openai",
  temperature: number = 0.3,
  model?: string
): Promise<z.infer<T>> {
  const { provider } = getProviderWithFailover(providerName);

  logger.debug("analysis.fact_extraction.custom.start", {
    provider: providerName,
  });

  const result = await provider.completeStructured(messages, schema, {
    model,
    temperature,
  });

  logger.info("analysis.fact_extraction.custom.complete", {
    provider: providerName,
  });

  return result as z.infer<T>;
}
